snap/validate.go

```diff
@@ -351,6 +351,44 @@ func validateProvenance(prov string) error {
 	return naming.ValidateProvenance(prov)
 }
 
+func validateGpioChardevSlots(info *Info) error {
+	// XXX: should this be relaxed to validate all types in case
+	// gpio-chardev slots are allowed for snap apps later?
+	if info.Type() != TypeGadget {
+		// not a gadget, nothing to do
+		return nil
+	}
+
+	chipLines := make(map[string][]string)
+	for _, slot := range info.Slots {
+		if slot.Interface != "gpio-chardev" {
+			continue
+		}
+		var sourceChip []string
+		if err := slot.Attr("source-chip", &sourceChip); err != nil {
+			return err
+		}
+		var lines string
+		if err := slot.Attr("lines", &lines); err != nil {
+			return err
+		}
+		for _, chip := range sourceChip {
+			chipLines[chip] = append(chipLines[chip], lines)
+		}
+	}
+
+	// validate lines for a chip across all gpio-chardev slots.
+	for chip, lines := range chipLines {
+		concat := strings.Join(lines, ",")
+		// ParseRange checks lines are valid and non-overlapping.
```

Comment:
shouldn't the check also be about overlap between different slots or is that something else?

Reply:
As @bboozzoo said, this first joins all lines for every chip over all slots, then calling ParseRange detects overlapping (and does some more validation). I will add a comment explaining the approach.

Reply:
@pedronis regarding the error, would it be better to specify which range span from which slot overlaps with what specifically? it would add a bit of complexity but definitely doable.

Reply:
yes, I think it would be better

Reply:
> can't we try to find all the clashes doing this

To make sure I understand, do you mean clashes per chip label across slots, right?

> or something for pairs of chips?

I am a little confused by this, would it be possible to give an example?

Reply:
yes, chip label across slots

sorry I was sloppy, what I meant with pairs of chips, is really the lines for the same chips in pairs of slots, but there's a lot of way to doing things, may main point is that the error should be clear about both lines, chips and slots involve in clashes

Reply:
Got it thanks, I will think of something